test/core/security/rbac_translator_test.cc

```diff
@@ -984,6 +1038,347 @@ TEST(GenerateRbacPoliciesTest, UnknownFieldInHeaders) {
+  EXPECT_THAT(rbacs.status().message(),
+              "\"audit_loggers[0].name\" is not a string.");
+}
+
+TEST_F(GenerateRbacPoliciesTest, IncorrectAuditLoggerIsOptionalType) {
+  const char* authz_policy =
+      "{"
+      "  \"name\": \"authz\","
+      "  \"allow_rules\": ["
+      "    {"
+      "      \"name\": \"allow_policy\""
+      "    }"
+      "  ],"
+      "  \"audit_logging_options\": {"
+      "    \"audit_loggers\": ["
+      "      {"
+      "        \"name\": \"test_logger\","
+      "        \"is_optional\": 123,"
+      "        \"config\": {}"
+      "      }"
+      "    ]"
+      "  }"
+      "}";
+  auto rbacs = GenerateRbacPolicies(authz_policy);
+  EXPECT_EQ(rbacs.status().code(), absl::StatusCode::kInvalidArgument);
+  EXPECT_THAT(rbacs.status().message(),
+              "\"audit_loggers[0].is_optional\" is not a boolean.");
+}
+
+TEST_F(GenerateRbacPoliciesTest, MissingAuditLoggerName) {
+  const char* authz_policy =
+      "{"
+      "  \"name\": \"authz\","
+      "  \"allow_rules\": ["
+      "    {"
+      "      \"name\": \"allow_policy\""
+      "    }"
+      "  ],"
+      "  \"audit_logging_options\": {"
+      "    \"audit_loggers\": ["
+      "      {"
+      "        \"config\": {}"
+      "      }"
+      "    ]"
+      "  }"
+      "}";
+  auto rbacs = GenerateRbacPolicies(authz_policy);
+  EXPECT_EQ(rbacs.status().code(), absl::StatusCode::kInvalidArgument);
+  EXPECT_THAT(rbacs.status().message(),
+              "\"audit_loggers[0].name\" is required.");
+}
+
+TEST_F(GenerateRbacPoliciesTest, MissingAuditLoggerConfig) {
+  const char* authz_policy =
+      "{"
+      "  \"name\": \"authz\","
+      "  \"allow_rules\": ["
+      "    {"
+      "      \"name\": \"allow_policy\""
+      "    }"
+      "  ],"
+      "  \"audit_logging_options\": {"
+      "    \"audit_loggers\": ["
+      "      {"
+      "        \"name\": \"test_logger\""
+      "      }"
+      "    ]"
+      "  }"
+      "}";
+  auto rbacs = GenerateRbacPolicies(authz_policy);
+  EXPECT_EQ(rbacs.status().code(), absl::StatusCode::kInvalidArgument);
+  EXPECT_THAT(rbacs.status().message(),
+              "\"audit_loggers[0].config\" is required.");
+}
+
+TEST_F(GenerateRbacPoliciesTest, UnsupportedAuditLogger) {
+  const char* authz_policy =
+      "{"
+      "  \"name\": \"authz\","
+      "  \"allow_rules\": ["
+      "    {"
+      "      \"name\": \"allow_policy\""
+      "    }"
+      "  ],"
+      "  \"audit_logging_options\": {"
+      "    \"audit_loggers\": ["
+      "      {"
+      "        \"name\": \"unknown_logger\","
+      "        \"config\": {}"
+      "      }"
+      "    ]"
+      "  }"
+      "}";
+  auto rbacs = GenerateRbacPolicies(authz_policy);
+  EXPECT_EQ(rbacs.status().code(), absl::StatusCode::kInvalidArgument);
+  EXPECT_THAT(rbacs.status().message(),
+              "\"audit_loggers[0].name\" unknown_logger is not supported.");
+}
+
```

Comment:
Can you add a test where you parse a non-empty `config`?
The `{"bad": true}` case parses it, but having a test where we pass a valid custom config and ensure that the values are accessible is worth it IMO

Reply:
This bad config test verifies that `{"bad":  true}` is passed into the parser appropriately, since `bad` is found by it. It also confirms that the translator can propagate the error from the parser properly. IMO, this should suffice as a unit test for the rbac translator.

Parsing valid config and returning a good logger is something the factory/registry unit test should do. I think we need to do that for our built-in loggers.

Reply:
What I was thinking is that the tests here just all ensure that the config is parsable, but this doesn't check that it correctly plumbs along those parsed values.

For example if there were a mistake in the code that made an empty config always get put on the returned object vs. this parsed config, would we have a test that caught that?

Reply:
I think I see your point now. I made the config store the json string and return it in `ToString()`. The two-logger test case is verifying that now.